include/cocaine/context/signal.hpp

```diff
@@ -21,72 +21,126 @@
 #ifndef COCAINE_CONTEXT_SIGNAL_HPP
 #define COCAINE_CONTEXT_SIGNAL_HPP
 
-#include "cocaine/tuple.hpp"
+#include "cocaine/locked_ptr.hpp"
+
+#include "cocaine/rpc/dispatch.hpp"
+#include "cocaine/rpc/frozen.hpp"
 
 #include <algorithm>
 #include <list>
-#include <mutex>
-
-#define BOOST_BIND_NO_PLACEHOLDERS
-#include <boost/signals2/signal.hpp>
 
 namespace cocaine {
 
-namespace signals = boost::signals2;
+template<class Tag> class retroactive_signal;
 
-template<
-    class Signature,
-    class IndexSequence = typename make_index_sequence<signals::signal<Signature>::arity>::type
->
-class retroactive_signal;
+namespace aux {
 
-template<typename... Args, size_t... Indices>
-class retroactive_signal<void(Args...), index_sequence<Indices...>> {
-    typedef signals::signal<void(Args...)> signal_type;
+template<class Event>
+struct async_visitor:
+    public boost::static_visitor<void>
+{
+    typedef typename io::basic_slot<Event>::tuple_type tuple_type;
 
-    // The actual underlying signal object.
-    signal_type wrapped;
+    async_visitor(const tuple_type& args_, asio::io_service& asio_):
+        args(args_),
+        asio(asio_)
+    { }
 
-    // Deferred signal arguments.
-    std::list<std::tuple<Args...>> mutable history;
-    std::mutex mutable mutex;
+    template<class Other>
+    result_type
+    operator()(const std::shared_ptr<io::basic_slot<Other>>& COCAINE_UNUSED_(slot)) const {
+        __builtin_unreachable();
+    }
 
-public:
-    typedef typename signal_type::slot_type slot_type;
+    result_type
+    operator()(const std::shared_ptr<io::basic_slot<Event>>& slot) const {
+        auto args = this->args;
 
-    auto
-    connect(const slot_type& slot) -> signals::connection {
-        try {
-            std::lock_guard<std::mutex> guard(mutex);
+        asio.post([slot, args]() mutable {
```

Comment:
I'd like to make sure that signals are always executed from the event loop, never from the signal caller's context (which might lead to some weird errors).